includes/options/class-amp-options-menu.php

```diff
@@ -1,54 +1,201 @@
+	 *
+	 * @since 0.6
+	 * @return array Supported post types list.
+	 */
+	public function get_supported_post_types() {
+		$core = get_post_types( array(
+			'name' => 'post',
+		), 'objects' );
+		$cpt  = get_post_types( array(
+			'public'   => true,
+			'_builtin' => false,
+		), 'objects' );
+
+		return array_merge( $core, $cpt );
 	}
 
-	// Helper function to avoid having the top-level menu as
-	// the first menu item
-	function remove_toplevel_menu_item() {
-		global $submenu;
-		if ( isset( $submenu['amp-plugin-options'][0] ) ) {
-			unset( $submenu['amp-plugin-options'][0] );
+	/**
+	 * Handle errors.
+	 *
+	 * @since 0.6
+	 */
+	public function errors() {
+		$on_update = (
+			isset( $_GET['settings-updated'] ) // WPCS: CSRF ok.
+			&&
+			true === (bool) wp_unslash( $_GET['settings-updated'] ) // WPCS: CSRF ok.
+		);
+
+		// Only apply on update.
+		if ( ! $on_update ) {
+			return;
+		}
+
+		$builtin_support = AMP_Post_Type_Support::get_builtin_supported_post_types();
+		$settings        = AMP_Options_Manager::get_option( 'supported_post_types', array() );
+		foreach ( $this->get_supported_post_types() as $post_type ) {
+			if ( ! isset( $post_type->name, $post_type->label ) || in_array( $post_type->name, $builtin_support, true ) ) {
+				continue;
+			}
+
+			$post_type_support = post_type_supports( $post_type->name, AMP_QUERY_VAR );
+			$value             = ! empty( $settings[ $post_type->name ] );
+
+			$error = null;
+			if ( true === $value && true !== $post_type_support ) {
+				/* translators: %s: Post type name. */
+				$error = __( '"%s" could not be activated because support is removed by a plugin or theme', 'amp' );
+			} elseif ( empty( $value ) && true === $post_type_support ) {
+				/* translators: %s: Post type name. */
+				$error = __( '"%s" could not be deactivated because support is added by a plugin or theme', 'amp' );
+			}
+
+			if ( isset( $error ) ) {
+				add_settings_error(
+					$post_type->name,
+					$post_type->name,
+					sprintf(
+						$error,
+						$post_type->label
+					)
+				);
+			}
 		}
 	}
+
+	/**
+	 * Post types support section renderer.
+	 *
+	 * @since 0.6
+	 */
+	public function render_post_types_support() {
+		$builtin_support = AMP_Post_Type_Support::get_builtin_supported_post_types();
+		?>
+		<fieldset>
+			<?php foreach ( $this->get_supported_post_types() as $post_type ) : ?>
+				<?php
+				$id         = AMP_Options_Manager::OPTION_NAME . "[supported_post_types][{$post_type->name}][]";
+				$is_builtin = in_array( $post_type->name, $builtin_support, true );
+				?>
+				<?php if ( $is_builtin ) : ?>
+					<input type="hidden" name="<?php echo esc_attr( $id ); ?>" value="1">
+				<?php endif; ?>
+				<input
+					type="checkbox"
+					value="1"
+					id="<?php echo esc_attr( $id ); ?>"
+					name="<?php echo esc_attr( $id ); ?>"
+					<?php checked( true, post_type_supports( $post_type->name, AMP_QUERY_VAR ) ); ?>
+					<?php disabled( $is_builtin ); ?>
```

Comment:
@ThierryA This is what I was talking about in regards to hard-coding a set of post types that are always active, and thus are disabled in the UI. Maybe we could instead just hide them altogether, but it turns out to be useful to have them present. The hidden input above ensures that there are some checked checkboxes which means we will get a non-empty array passed in to validate.